pageserver/src/remote_storage/storage_sync/index.rs

```diff
@@ -0,0 +1,378 @@
+//! In-memory index to track the timeline files in the remote strorage's archives.
+//! Able to restore itself from the storage archive data and reconstruct archive indices on demand.
```

Comment:
No real estimates, but I expect this to be inevitably slow (we don't really delete anything from S3, only accumulate now), so the full restoration has to be done in the sync thread, letting the rest of the pageserver run meanwhile.

Yet, the way I see it, before running the rest of the pageserver, we have to get to know _something_ about the remote storage contents: otherwise some local timelines could be outdated or even diverging with the remote version, we have to disallow backing those up IMO, by presumably offloading(is it a good idea?) such timelines from pageserver.

To do so, I see three ways to improve the current situation:
* collect the local timelines, do multiple "s3 list" commands with their prefixes to get a limited set of file listings from S3 to understand, whether the current local timeline is out of data or not.
I've planned comparing `disk_consistent_lsn` from the remote archive names and local metadata for that, but now not so sure due to your comments.

* keep the index map (`HashMap<(ZTenantId, ZTimelineId), RemoteTimeline>`) synced periodically to S3, so that on restoration we would need to read one file and then later compare it with the real contents and update, if needed.

Both ways complement each other, but still require eventually listing full S3 contents at some point.
So far looks good enough to start since we can postpone that potentially call into a separate thread, but I have no good plans on how to mitigate that in the future, besides actually removing the files from S3.

Reply:
Yes, now we rely on the incremental nature of all files and wait for the new file format to arrive.
I've heard polar opinions about that (100% incremental vs every file could be volatile), so not sure if it's wise to add those checks right now.

Also, there are plans to add checksum verifications eventually, that could be a good moment to include these checks.

For now, no explicit checks are done, but overwrites may come from archive extraction only with the current logic:
* checkpoint uploads are done for different `disk_consistent_lsn`s only, so archive names don't overlap. There's currently already a `already_uploaded_files` collection in `upload.rs` in this PR to ensure, that nothing gets extracted over, if the file path is in the local FS.
This does not consider file checksums or consistency in any other way, but it's a separate big story.

* metadata overwrites during timeline downloads: inevitable, but can be reduced to a single overwrite. Right now checkpoint archives are unpacked from the earliest `disk_consistent_lsn` to the latest, with every archive overwriting the metadata file.


I don't do incoming checkpoint files verification, since the `already_uploaded_files` logic should filter such files off, but can do.
Otherwise, we seem to be ok with files overwrites for now?

Reply:
> (100% incremental vs every file could be volatile)

Let me clarify this, it shouldn't depend much on a file format. I think about checks like if we have timeline locally with some data then forbid extraction of a remote timeline into the same directory. If local timeline is active we might do something the underlying code doesnt take into account. Extraction is also not atomic, so when pageserver tries to find a page and sees half extracted checkpoint it can miss something (please correct me if I'm wrong, I might've missed something). Can we overwrite local metadata with newer values (e.g lsns) with older remote metadata? It is hard to tell what will happen then. So that's why I'm raising this concern

> we seem to be ok with files overwrites for now?

It is just a bit scary for me to imagine multiple pageservers writing checkpoints to the same underlying s3 storage. Even if it works I think that this is not the right thing to do

Reply:
> Can we overwrite local metadata with newer values (e.g lsns) with older remote metadata

Not with the current approach at least: since the download is done once max and we check the initial local and remote states before scheduling the tasks.

> multiple pageservers writing checkpoints to the same underlying s3 storage

Yes, this is a valid concern and we have nothing in the code to protect from that, even though we postulate in the docs, that only a single pageserver has to have an access for a certain bucket.
We have plans to adjust the s3 code to use prefixes inside the bucket to share the same bucket with multiple pageservers, but that's a later plan.

So far it is dangerous indeed, but not intended to be fixed in this PR.

Reply:
> since the download is done once max and we check the initial local and remote states before scheduling the tasks.

Could you elaborate? We consider metadata only of a last downloaded checkpoint, right? This is not possible right now but when we will have attach/detach routines can we have a case when there is an active timeline and we extract some old remote checkpoint into it. Will we overwrite local metadata?

> So far it is dangerous indeed, but not intended to be fixed in this PR.

Agree. Lets have it separately. See https://github.com/zenithdb/zenith/issues/971. Maybe turn it to an epic itself or just split other issues from it

Reply:
> Will we overwrite local metadata?

Yes, right now we will, so there is a concern, good that we have a separate issue for that now.